lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php

```diff
@@ -263,41 +287,76 @@ private function getEntity(array $data, $dqlAlias)
         }
 
         if (isset($this->_hints[Query::HINT_REFRESH_ENTITY]) && isset($this->rootAliases[$dqlAlias])) {
-            $this->registerManaged($this->_metadataCache[$className], $this->_hints[Query::HINT_REFRESH_ENTITY], $data);
+            $this->registerManaged($this->_metadataCache->{$className}, $this->_hints[Query::HINT_REFRESH_ENTITY], $data);
         }
 
         $this->_hints['fetchAlias'] = $dqlAlias;
 
-        return $this->_uow->createEntity($className, $data, $this->_hints);
+        $managedEntity = ($this->entityTryGetCallbacksByEntityName->{$className})($data);
+
+        // If the entity is not existing in the UnitOfWork, then we
+        // are the one creating it: let's track the object hash to
+        // allow writing to it. Note that we discard `$managedEntity`
+        // anyway, as we need to call `createEntity` with the data
+        // in any case to allow refreshing proxy information, for
+        // example
+        if (! $managedEntity) {
+            $entity = $this->_uow->createEntity($className, $data, $this->_hints);
+
+            $this->createdEntities[\spl_object_hash($entity)] = true;
+
+            return $entity;
+        }
+
+        $entity = $this->_uow->createEntity($className, $data, $this->_hints);
+
+        // If the entity is not created by us, we can only consider
+        // it to be created in here if it's a non-initialized proxy
+        if ($entity instanceof Proxy && ! $entity->__isInitialized()) {
+            $this->createdEntities[\spl_object_hash($entity)] = true;
+        }
+
+        return $entity;
     }
 
     /**
+     * @internal do not use this or you shall be punished with the wrath of the debugger
+     * @private
+     *
      * @param string $className
-     * @param array  $data
      *
-     * @return mixed
+     * @return callable
      */
-    private function getEntityFromIdentityMap($className, array $data)
```

Comment:
Super glad that this is gone: we actually gained quite some performance here :+1: